dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamodb
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+)
+
+const (
+	typeString         = "String"
+	typeBool           = "Bool"
+	typeNumberInt      = "NumberInt"
+	typeNumberFloat    = "NumberFloat"
+	typeBinary         = "Binary"
+	typeList           = "List"
+	typeMap            = "Map"
+	typeStringSet      = "StringSet"
+	typeNumberIntSet   = "NumberIntSet"
+	typeNumberFloatSet = "NumberFloatSet"
+	typeBinarySet      = "BinarySet"
+
+	errThreshold      = float64(0.001)
+	conflictThreshold = float64(0.05)
+)
+
+type dynamoClient interface {
+	ListTables(input *dynamodb.ListTablesInput) (*dynamodb.ListTablesOutput, error)
+	DescribeTable(input *dynamodb.DescribeTableInput) (*dynamodb.DescribeTableOutput, error)
+	Scan(input *dynamodb.ScanInput) (*dynamodb.ScanOutput, error)
+}
+
+// ProcessSchema performs schema conversion for source tables in a DynamoDB
+// database. We use the standard APIs to obtain source table's schema
+// information, including primary keys and secondary indexes. DynamoDB is a
+// schemaless database that some optional attributes can be missed or has
+// different data types in some rows. Therefore, we have to sample a number of
```

Comment:
Suggest rephrasing: I think the comment is a bit too detailed and is likely to get out-of-date.

How about:

ProcessSchema performs schema conversion for source tables in a DynamoDB
database. Since DynamoDB is a schemaless database, this process is imprecise.
We obtain schema information from two sources: from the table's metadata,
and from analyzing a sample of the table's rows.

Reply:
Fixed.